lib/terraform/runner.rb

```diff
@@ -15,83 +15,53 @@ def available?
+      #   4. ActionType::RETRIEVE (Fetch)       fetch a existing stack object json from terraform-runner.
+      #   5. ActionType::CANCEL   (Stop)        stop running stack job in terraform-runner.
+      # * keys allowed in ':options'
+      #   - :input_vars                  [Hash]   (optional) key/value pairs, as input variables for the terraform-runner run job.
+      #   - :input_vars_type_constraints [Hash]   (optional) key/value(type constraints object, from Terraform Runner) pairs.
+      #   - :tags                        [Hash]   (optional) key/value pairs tags for terraform-runner Provisioned resources.
+      #   - :credentials                 [Array]  (optional) list of authentication objects for the terraform run job.
+      #   - :env_vars                    [Hash]   (optional) key/value pairs, used as environment variables, for terraform-runner run job.
+      #   - :name                        [String] (optional) name for new created stack in terraform-runner.
+      #   - :stack_id                    [String] [required] if Reconfigure/Retire/Retrieve/Cancel actions.
       #
-      # @return [Terraform::Runner::ResponseAsync] Response object of terraform-runner api call
-      def delete_stack(stack_id, template_path, input_vars: {}, input_vars_type_constraints: [], credentials: [], env_vars: {})
-        _log.debug("Run_aysnc/delete_stack('#{stack_id}') for template: #{template_path}")
-        response = run_terraform_runner_stack_api(
-          Request.new(ActionType::DELETE)
-            .stack_id(stack_id)
-            .template_path(template_path)
-            .credentials(credentials)
-            .input_vars(input_vars, input_vars_type_constraints)
-            .tenant_id(stack_tenant_id)
-            .env_vars(env_vars)
-        )
-        Terraform::Runner::ResponseAsync.new(response.stack_id)
+      def run(action_type, template_path, options = {}, _stack_id = nil)
+        _log.debug("Run #{action_type} for template: #{template_path}")
+
+        case action_type
+        when ActionType::CREATE, ActionType::UPDATE, ActionType::DELETE, ActionType::RETRIEVE, ActionType::CANCEL
+          response = run_terraform_runner_stack_api(
+            Request.new(action_type)
+              .template_path(template_path)
+              .credentials(options[:credentials])
+              .input_vars(options[:input_vars], options[:input_vars_type_constraints])
+              .tenant_id(stack_tenant_id)
+              .tags(options[:tags])
+              .env_vars(options[:env_vars])
+              .name(options[:name])
+              .stack_id(options[:stack_id])
+          )
+
+          Terraform::Runner::ResponseAsync.new(response.stack_id)
+        when ActionType::TEMPLATE_VARIABLES
+          raise "Not supported action type in this method, instead use method parse_terraform_variables"
+        else
+          raise "Not supported action type '#{action_type}'"
+        end
```

Comment:
I like to handle these failure cases early in the method versus having the entire thing indended under a case/conditional.  It reads a little better, getting all of the checks out of the way early and the rest of the method assumes you've made it past the gatekeeping.

You could e.g.:
```
raise "Not supported action type in this method, instead use method parse_terraform_variables" if action_type == ActionType::TEMPLATE_VARIABLES
raise "Not supported action type '#{action_type}'" unless ActionType.actions.include?(action_type)

response = run_terraform_runner_stack_api(
etc...
```